### ISSUE TITLE
most os_* moduels don't use clouds.yaml verify:false setting when validate_certs is not provided.

### ISSUE THREAD

**Issue by rasselin** [ansible/ansible-modules-core#3981]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

os_network
os_subnet
os_router
os_image
os_security_group
os_security_group_rule
os_keypair
os_server
_maybe others_
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = 
  configured module search path = Default w/o overrides

```
##### OS / ENVIRONMENT

ubuntu 14.04
##### SUMMARY

When creating any of these modules and omitting the validate_certs: option,
but setting the cloud: option & including verify: false in clouds.yaml
the modules fail with SSL error such as
Inner Exception: SSL exception connecting to https://10.245.78.7:5000/v2.0/tokens: [Errno 1] _ssl.c:510: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed)"
##### STEPS TO REPRODUCE

```
1. Setup a private/dev cloud with self-signed certs.
2. Setup ~/.config/openstack/clouds.yaml cloud configuration
3. Ensure that openstack server list command fails when verify = true is not in the cloud configuration to show that ssl cert is not valid, but succeeds when verify = false

4. Using this external ansible role: https://github.com/openstack/ansible-role-cloud-launcher/blob/master/resources.yml.sample which is a simple wrapper to execute all these os-* ansible modules.

Observe that they fail in turn, but when setting validate_certs: False to the failing os_* modules, they succeed.

```

##### EXPECTED RESULTS

Expecting the various os_\* ansible modules to successfully provision and interact with the configured cloud. 
##### ACTUAL RESULTS

os_ ansible modules fail due to invalid certs.

```

TASK [ansible-role-cloud-launcher : Create profiles networks for gpc-asselin] **
task path: /etc/ansible/roles/ansible-role-cloud-launcher/tasks/create_clouds_resources.yml:93
failed: [localhost] (item=({u'routers': [{u'interfaces': u'default-subnet', u'name': u'default-router', u'network': u'ext-net'}], u'security_groups_rules': [{u'security_group': u'webserver', u'protocol': u'tcp', u'port_range_max': 80, u'port_range_min': 80, u'remote_ip_prefix': u'0.0.0.0/0'}, {u'security_group': u'webserver', u'protocol': u'tcp', u'port_range_max': 443, u'port_range_min': 443, u'remote_ip_prefix': u'0.0.0.0/0'}], u'name': u'asselin-gpc', u'security_groups': [{u'name': u'webserver', u'description': u'Allow HTTP/HTTPS traffic'}], u'servers': [{u'name': u'bs01', u'key_name': u'bootstrap-key', u'image': u'ubuntu-trusty', u'volumes': [{u'name': u'gfs01', u'volume_type': u'3par-vol', u'size': 500}], u'flavor': u'm1.large', u'security_groups': u'webserver', u'network': u'default-network'}, {u'name': u'bs02', u'key_name': u'bootstrap-key', u'image': u'ubuntu-trusty', u'volumes': [{u'name': u'gfs02', u'volume_type': u'3par-vol', u'size': 500}], u'flavor': u'm1.large', u'security_groups': u'webserver', u'network': u'default-network'}]}, {u'name': u'default-network'})) => {"failed": true, "item": [{"name": "asselin-gpc", "routers": [{"interfaces": "default-subnet", "name": "default-router", "network": "ext-net"}], "security_groups": [{"description": "Allow HTTP/HTTPS traffic", "name": "webserver"}], "security_groups_rules": [{"port_range_max": 80, "port_range_min": 80, "protocol": "tcp", "remote_ip_prefix": "0.0.0.0/0", "security_group": "webserver"}, {"port_range_max": 443, "port_range_min": 443, "protocol": "tcp", "remote_ip_prefix": "0.0.0.0/0", "security_group": "webserver"}], "servers": [{"flavor": "m1.large", "image": "ubuntu-trusty", "key_name": "bootstrap-key", "name": "bs01", "network": "default-network", "security_groups": "webserver", "volumes": [{"name": "gfs01", "size": 500, "volume_type": "3par-vol"}]}, {"flavor": "m1.large", "image": "ubuntu-trusty", "key_name": "bootstrap-key", "name": "bs02", "network": "default-network", "security_groups": "webserver", "volumes": [{"name": "gfs02", "size": 500, "volume_type": "3par-vol"}]}]}, {"name": "default-network"}], "msg": "Error fetching network list: SSL exception connecting to https://10.245.78.7:5000/v2.0/tokens: [Errno 1] _ssl.c:510: error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed"}



```

**Comment by rasselin** [comment]
@rcarrillocruz  submitted per your request

**Comment by rcarrillocruz** [comment]
@rasselin can you confirm this is seen even with verify: false on your clouds.yaml stanza?

I believe you told me affirmative on IRC, but better put it explicitly here for the record.

**Comment by rasselin** [comment]
Yes, here's a copy of my clouds.yaml with password sanitized:

```
clouds:
  gpc-asselin:
    auth:
      auth_url: 'https://10.245.78.7:5000/v2.0'
      username: 'rasselin'
      password: 'secret'
      project_name: 'rasselin'
    region_name: region1
    verify: false
```

**Comment by ansibot** [comment]
@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by juliakreger** [comment]
It appears that the issue may be rooted in keystoneauth, however further investigation is required.

**Comment by juliakreger** [comment]
We are going to need a little more information, specifically the versions of the various openstack client libraries as well as shade, os-client-config, keystoneauth.  This can be retrieved by executing `pip list`.  This information is needed in order to trace down the issue since multiple libraries are involved.

needs_info

**Comment by rasselin** [comment]
Hi Julia, You should be able to reproduce with the pip openstack libraries installed by ansible's requirements.

Either way here is what is installed in my ansible venv:

> ansible (2.1.0.0)
> appdirs (1.4.0)
> Babel (2.3.4)
> cffi (1.6.0)
> cliff (2.1.0)
> cmd2 (0.6.8)
> cryptography (1.4)
> debtcollector (1.5.0)
> decorator (4.0.10)
> dogpile.cache (0.6.1)
> enum34 (1.1.6)
> funcsigs (1.0.2)
> functools32 (3.2.3.post2)
> futures (3.0.5)
> httplib2 (0.9.2)
> idna (2.1)
> ipaddress (1.0.16)
> iso8601 (0.1.11)
> Jinja2 (2.8)
> jsonpatch (1.14)
> jsonpointer (1.10)
> jsonschema (2.5.1)
> keystoneauth1 (2.8.0)
> MarkupSafe (0.23)
> monotonic (1.1)
> msgpack-python (0.4.7)
> munch (2.0.4)
> netaddr (0.7.18)
> netifaces (0.10.4)
> openstacksdk (0.8.6)
> os-client-config (1.18.0)
> oslo.config (3.11.0)
> oslo.i18n (3.7.0)
> oslo.serialization (2.9.0)
> oslo.utils (3.13.0)
> paramiko (2.0.0)
> pbr (1.10.0)
> pip (7.1.2)
> positional (1.1.0)
> prettytable (0.7.2)
> pyasn1 (0.1.9)
> pycparser (2.14)
> pycrypto (2.6.1)
> pyparsing (2.1.5)
> python-cinderclient (1.8.0)
> python-designateclient (2.1.0)
> python-glanceclient (2.0.0)
> python-heatclient (1.2.0)
> python-ironicclient (1.3.1)
> python-keystoneclient (3.1.0)
> python-magnumclient (2.1.0)
> python-neutronclient (4.2.0)
> python-novaclient (4.1.0)
> python-openstackclient (2.6.0)
> python-swiftclient (3.0.0)
> python-troveclient (2.2.0)
> pytz (2016.4)
> PyYAML (3.11)
> requests (2.10.0)
> requestsexceptions (1.1.3)
> setuptools (18.2)
> shade (1.8.1.dev11)
> simplejson (3.8.2)
> six (1.10.0)
> stevedore (1.15.0)
> unicodecsv (0.14.1)
> warlock (1.2.0)
> wheel (0.24.0)
> wrapt (1.10.8)

**Comment by emonty** [comment]
I believe I see the issue ... in openstack_full_argument_spec we have this:
        verify=dict(default=True, type='bool', aliases=['validate_certs']),

Which means we're setting a default value ansible-side regardless of whether the user said anything or not. Since we pass in values, they override.

I'll try to come up with an answer - none of the ones I have so far are "good" :)

**Comment by ansibot** [comment]
@rasselin ping, this issue is still waiting on your feedback. We will close the issue if you do not respond.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by rasselin** [comment]
!needs_info
I'm not sure why ansibot added needs_info. Anyway, if there's no 'good' solution, the workaround is fairly straightforward. Just add the validate_certs on each os_\* module used. For example in ansible-role-cloud-launcher https://review.openstack.org/#/c/360675/
I wouldn't object to closing won't fix.

**Comment by emonty** [comment]
https://review.openstack.org/403748 is step one in fixing this. PR will be coming for the ansible side.

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by daledude** [comment]
Any hope of getting this pushed? Using validate_certs everywhere is painful.
